File: packages/guards/src/refund-policy-guard.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RefundPolicyGuardOptions {
  action: 'block' | 'warn';
}

const TRANSACTION_PATTERNS = [
  /\bpurchase\b/i,
  /\border\b/i,
  /\bsubscrib/i,
  /\bpayment\b/i,
  /\bcharge\b/i,
  /\bbuy\b/i,
  /\bcheckout\b/i,
];

const CONSUMER_PROTECTION = [
  /\brefund\b/i,
  /\breturn\s+policy\b/i,
  /\bcancell?ation\b/i,
  /\bcooling[- ]off\s+period\b/i,
  /\bmoney[- ]?back\b/i,
  /\bcancel\s+anytime\b/i,
  /\bcancel\s+subscription\b/i,
  /\bopt\s+out\b/i,
];

export function refundPolicyGuard(options: RefundPolicyGuardOptions): Guard {
  return {
    name: 'refund-policy-guard',
    version: '0.1.0',
    description: 'Ensure agentic transactions include refund/cancellation rights',
    category: 'compliance',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();

      const hasTransaction = TRANSACTION_PATTERNS.some((p) => p.test(text));
      const hasProtection = CONSUMER_PROTECTION.some((p) => p.test(text));

      const triggered = hasTransaction && !hasProtection;
      return {
        guardName: 'refund-policy-guard',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { reason: 'Transaction detected without consumer protection information' }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/refund_policy_guard.py
"""Ensure agentic transactions include refund/cancellation rights."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_TRANSACTION_PATTERNS = [
    re.compile(r"\bpurchase\b", re.IGNORECASE),
    re.compile(r"\border\b", re.IGNORECASE),
    re.compile(r"\bsubscrib", re.IGNORECASE),
    re.compile(r"\bpayment\b", re.IGNORECASE),
    re.compile(r"\bcharge\b", re.IGNORECASE),
    re.compile(r"\bbuy\b", re.IGNORECASE),
    re.compile(r"\bcheckout\b", re.IGNORECASE),
]

_CONSUMER_PROTECTION = [
    re.compile(r"\brefund\b", re.IGNORECASE),
    re.compile(r"\breturn\s+policy\b", re.IGNORECASE),
    re.compile(r"\bcancell?ation\b", re.IGNORECASE),
    re.compile(r"\bcooling[- ]off\s+period\b", re.IGNORECASE),
    re.compile(r"\bmoney[- ]?back\b", re.IGNORECASE),
    re.compile(r"\bcancel\s+anytime\b", re.IGNORECASE),
    re.compile(r"\bcancel\s+subscription\b", re.IGNORECASE),
    re.compile(r"\bopt\s+out\b", re.IGNORECASE),
]


class _RefundPolicyGuard:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "refund-policy-guard"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()

        has_transaction = any(p.search(text) for p in _TRANSACTION_PATTERNS)
        has_protection = any(p.search(text) for p in _CONSUMER_PROTECTION)

        triggered = has_transaction and not has_protection
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="refund-policy-guard",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="Transaction detected without consumer protection information" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"reason": "Transaction detected without consumer protection information"} if triggered else None,
        )


def refund_policy_guard(*, action: str = "block") -> _RefundPolicyGuard:
    return _RefundPolicyGuard(action=action)
